profile/label_nodes.go

```diff
@@ -0,0 +1,97 @@
+package profile
+
+// AddLabelNodes adds psuedo-stack-frames "label:value" to each Sample with
+// labels matching the supplied keys.
+//
+// rootKeys adds frames at the root of the callgraph (first key becomes new root).
+// leafKeys adds frames at the leaf of the callgraph (last key becomes new leaf).
+//
+// Returns whether there were matches found for the label keys.
+func (p *Profile) AddLabelNodes(rootKeys, leafKeys []string) (rootm, leafm bool) {
+	// Find where to insert the new locations and functions at the end of
+	// their ID spaces.
+	var maxLocID uint64
+	var maxFunctionID uint64
+	for _, loc := range p.Location {
+		if loc.ID > maxLocID {
+			maxLocID = loc.ID
+		}
+	}
+	for _, f := range p.Function {
+		if f.ID > maxFunctionID {
+			maxFunctionID = f.ID
+		}
+	}
+	nextLocID := maxLocID + 1
+	nextFuncID := maxFunctionID + 1
+
+	// Intern the new frames we are generating.
+	locsByName := map[string]*Location{}
+	functionsByName := map[string]*Function{}
+
+	internFunction := func(functionName string) *Function {
```

Comment:
Maybe internFunction and internLocation could be a single intern() function? Both are interned at the same time and the function name is the key.

Reply:
Yeah, I considered it, but I think there's an advantage to having them separate: you can early-return the found interned item. I personally think it's a little clearer with the early return, which wouldn't be possible if they were combined. Happy to merge them if you feel any stronger than that.